### PR TITLE
Exclude slow tests by stripping tags from tests

### PR DESCRIPTION
**Pull request by jiegillet**
In the spirit of discussing over code rather than hypotheticals, here is my implementation of stripping `@tag :pending` lines from the test themselves and using `--exclude slow`.

### CODE REVIEW

File: test/hello-world/test/hello_world_test.exs [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
-  test("failing test") do
+  @tag :pending
+  test "failing test" do
     assert [1, 2, 3] == [2, 3]
   end
+  @tag task_id: 3
+  @tag :pending
+  @tag :slow
+  test "Would fail but should not be called from the test runner" do
+    assert "tim" == "angelika"
```
Comment (angelikatyborska) [inline test/hello-world/test/hello_world_test.exs:17]: 😂

File: exercism_test_helper/lib/test_source/transformer.ex [hunk 2]
```diff
@@ -46,4 +47,12 @@ defmodule TestSource.Transformer do
...
     end
   end
+
+  defp remove_pending_tag({name, meta, args}) when is_list(args),
+    do: {name, meta, Enum.reject(args, &pending_tag?/1)}
+
+  defp remove_pending_tag(node), do: node
+
+  defp pending_tag?({:@, _, [{:tag, _, [:pending]}]}), do: true
+  defp pending_tag?(node), do: false
```
Comment (angelikatyborska) [inline exercism_test_helper/lib/test_source/transformer.ex:57]: Compiler warning:
```suggestion
  defp pending_tag?(_node), do: false
```

File: exercism_test_helper/lib/test_source/transformer.ex [hunk 3]
```diff
@@ -46,4 +47,12 @@ defmodule TestSource.Transformer do
       IO.puts("[test started] test " <> unquote(name))
     end
   end
+
+  defp remove_pending_tag({name, meta, args}) when is_list(args),
+    do: {name, meta, Enum.reject(args, &pending_tag?/1)}
+
+  defp remove_pending_tag(node), do: node
+
+  defp pending_tag?({:@, _, [{:tag, _, [:pending]}]}), do: true
```
Comment (angelikatyborska) [inline exercism_test_helper/lib/test_source/transformer.ex:56]: I _think_ `ExUnit.configure(exclude: :pending, trace: true)` will also exclude tests where the `:pending` tag was defined with some value (also value different than `true`), e.g. `@tag pending: 3`. For completeness, we should strip that too. I would suggest adding a test in `transformer_test.exs` like this (currently failing):

```elixir
  describe "testsuite - more tests" do
    test "strips pending tags" do
      defsuite do
        test "will be run 1" do
          assert(true)
        end

        @tag :pending
        test "will be run 2" do
          assert(true)
        end

        @tag pending: true
        test "will be run 3" do
          assert(true)
        end

        @tag :pending
        @tag :slow
        test "will not be run 1" do
          assert(true)
        end

        @tag :slow
        test "will not be run 2" do
          assert(true)
        end
      end

      desired = """
      [test started] test will be run 1
      [test started] test will be run 2
      [test started] test will be run 3
      """

      # add exclude: :pending to mimic the test_helpers.exs files in every exercise
      output = capture_io(fn -> run([exclude: [:slow, :pending]]) end)

      assert output == desired
    end
  end
```

**Review by angelikatyborska — APPROVED** [review]
I think this is the best approach to this problem given the limitations of `--include`/`--exclude` flags for the `mix task` and the different `test_helpers.exs` that some exercises might have 👍

File: exercism_test_helper/lib/test_source/transformer.ex (hunk 2 above)
Comment (jiegillet) [reply exercism_test_helper/lib/test_source/transformer.ex]: Thank you

**Comment by neenjaw** [comment]
Feel free to merge if you think it's the best approach

**Comment by jiegillet** [comment]
I don't know why the test is still failing, I changed the expected result to match the new one and the test passes on my machine.

**Comment by angelikatyborska** [comment]
Perfect 🙂 Now we only need to add the tags to the appropriate test cases.

**Comment by jiegillet** [comment]
I'm working on it :)

**Pull request merged**